File: R/naive_bayes_classifier.R
naive_bayes_classifier <- function(X, y) {
  classes <- unique(y)
  class_probabilities <- table(y) / length(y)
  feature_probabilities <- list()
  
  # Calculate feature probabilities
  for (cls in classes) {
    cls_indices <- which(y == cls)
    feature_probabilities[[cls]] <- apply(X[cls_indices, ], 2, table) / length(cls_indices)
  }
  
  list(
    predict = function(X_new) {
      predictions <- sapply(1:nrow(X_new), function(i) {
        class_scores <- sapply(classes, function(cls) {
          class_score <- class_probabilities[cls]
          for (j in 1:ncol(X_new)) {
            class_score <- class_score * feature_probabilities[[cls]][[j]][X_new[i, j]]
          }
          class_score
        })
        names(which.max(class_scores))
      })
      predictions
    }
  )
}

# Example usage
X <- matrix(c(1, 0, 1, 1, 0, 1, 0, 0), ncol = 2, byrow = TRUE)
y <- c("Spam", "Spam", "Ham", "Ham")
classifier <- naive_bayes_classifier(X, y)
print("Predictions:")
print(classifier$predict(X))

File: Go/naive_bayes_classifier.go
package main

import (
    "fmt"
)

type NaiveBayesClassifier struct {
    classes          []string
    classProb        map[string]float64
    featureProb      map[string]map[int]map[int]float64
}

func (nb *NaiveBayesClassifier) Fit(X [][]int, y []string) {
    totalSamples := len(y)
    nb.classProb = make(map[string]float64)
    nb.featureProb = make(map[string]map[int]map[int]float64)

    for _, cls := range y {
        nb.classProb[cls]++
    }
    for cls := range nb.classProb {
        nb.classProb[cls] /= float64(totalSamples)
        nb.featureProb[cls] = make(map[int]map[int]float64)
    }

    for i, cls := range y {
        for j, feature := range X[i] {
            if nb.featureProb[cls][j] == nil {
                nb.featureProb[cls][j] = make(map[int]float64)
            }
            nb.featureProb[cls][j][feature]++
        }
    }
    for cls := range nb.featureProb {
        for j := range nb.featureProb[cls] {
            total := float64(totalSamples)
            for feature := range nb.featureProb[cls][j] {
                nb.featureProb[cls][j][feature] /= total
            }
        }
    }
}

func main() {
    X := [][]int{{1, 0}, {1, 1}, {0, 1}, {0, 0}}
    y := []string{"Spam", "Spam", "Ham", "Ham"}
    classifier := NaiveBayesClassifier{}
    classifier.Fit(X, y)
    fmt.Println("Naive Bayes Classifier in Go")
}
